types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+				" is already on this surface");
+			return;
+		}
+	}
+
+	constraint = calloc(1, sizeof(struct wlr_pointer_constraint_v1));
+	if (constraint == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	constraint->resource = resource;
+	constraint->surface = surface;
+	constraint->seat_client = seat_client;
+	constraint->current.region = region;
+	constraint->pending.region = region;
+	constraint->lifetime = lifetime;
+	constraint->type = type;
+	constraint->active = false;
+	constraint->pointer_constraints = pointer_constraints;
+
+	constraint->surface_destroy.notify = handle_surface_destroy;
+	wl_signal_add(&surface->events.destroy, &constraint->surface_destroy);
+	constraint->surface_commit.notify = handle_surface_commit;
+	wl_signal_add(&surface->events.commit, &constraint->surface_commit);
+
+	wl_resource_set_implementation(constraint->resource,
+		locked_pointer ? (void*)&locked_pointer_impl : (void*)&confined_pointer_impl,
+		constraint, pointer_constraint_destroy_resource);
+
+	wlr_log(L_DEBUG, "new %s_pointer %p (res %p)",
+		locked_pointer ? "locked" : "confined", constraint, constraint->resource);
+
+	wl_list_insert(list, &constraint->link);
+
+	if (seat_client->seat->pointer_state.focused_surface == surface) {
+		assert(!seat_client->seat->current_pointer_constraint);
+		seat_client->seat->current_pointer_constraint = constraint;
+	}
+
+	wlr_signal_emit_safe(&pointer_constraints->events.request_constraint,
+		constraint);
+}
+
+static void pointer_constraints_lock_pointer(struct wl_client *client,
+		struct wl_resource *cons_resource, uint32_t id,
+		struct wl_resource *surface, struct wl_resource *pointer,
+		struct wl_resource *region, uint32_t lifetime) {
+	pointer_constraint_create(client, cons_resource, id, surface, pointer,
+		region, lifetime, WLR_POINTER_CONSTRAINT_V1_LOCKED);
+}
+
+static void pointer_constraints_confine_pointer(struct wl_client *client,
+		struct wl_resource *cons_resource, uint32_t id,
+		struct wl_resource *surface, struct wl_resource *pointer,
+		struct wl_resource *region, uint32_t lifetime) {
+	pointer_constraint_create(client, cons_resource, id, surface, pointer,
+		region, lifetime, WLR_POINTER_CONSTRAINT_V1_CONFINED);
+}
+
+static const struct zwp_pointer_constraints_v1_interface pointer_constraints_impl = {
+	.destroy         = resource_destroy,
+	.lock_pointer    = pointer_constraints_lock_pointer,
+	.confine_pointer = pointer_constraints_confine_pointer,
+};
+
+static void pointer_constraints_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_pointer_constraints_v1 *cons = data;
+	assert(client && cons);
+
+	struct wl_resource *resource = wl_resource_create(client,
+			&zwp_pointer_constraints_v1_interface, version, id);
+	if (resource == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(resource, &pointer_constraints_impl,
+		cons, NULL);
```

Comment:
Should put the resource in a list in `wlr_pointer_constraints_v1` (see `wl_resource_get_link`)